### PR TITLE
Update clang-format to version 5.0

### PR DESCRIPTION
**Pull request by Rombur**
**edited by @dalg24**
This PR, in its original form, introduced massive changes in our base image.
I reworked it and took out the changes related to installing a new MPI with ABI compatibility because I would like to try it out before we merge into master.
I also postponed the install of an alternate clang version.  My understanding is that @Rombur wanted to test it out in conjunction with CUDA 9.0 but it the meantime we ran into compile issues with Intrepid2 (c.f. #308) with the new version of CUDA.

- Open MPI upgrade to be able to call mpirun outside the container [REMOVED]
- LLVM/Clang/OpenMP runtime upgrade to 5.0.0
- new code formatting enforced by clang-format-5.0
- system package manager install of clang-3.9 [REMOVED]
- boost upgrade
- trilinos and kokkos-tools updates [CHANGED]
- new versions of PETSc and libMesh [ADDED]

NOTE: All PRs will need to be rebased because of the clang-format update to 5.0 and you will want to reset you git pre-commit hooks.

### CODE REVIEW

**Comment by dalg24** [comment]
Related to #302

**Comment by Rombur** [comment]
We will need to change the tag in jenkins matrix when we merge this.

**Comment by dalg24** [comment]
Bruno please update description when this is ready for review.  There are a lot of changes happening in this PR and it would be easy to overlook some of them.
- [ ] openmpi upgrade to be able to call mpirun outside the container
- [ ] clang+openmp upgrade
- [ ] clang-format upgrade
- [ ] system package manager install of clang-3.9
- [ ] boost upgrade
- [ ] trilinos and kokkos updates

**Comment by dalg24** [comment]
BTW, do we need to update our .clang-format to use the "ReflownComments" you mentioned in #302?
I looked at the logs on Travis and they essentially are changes in indentations/line breaks.

**Comment by Rombur** [comment]
It's enabled by default ```Option ReflowComments added and enabled by default```

File: docker/Dockerfile_stack [hunk 1]
```diff
@@ -221,8 +243,8 @@ ENV LIBMESH_DIR=/opt/libmesh/1.2.0
 
 # download Trilinos
 # git clone --depth 1 only works for the last commit -> use a tarball instead
-RUN export TRILINOS_HASH=4c8adc92fcbad83b827f9f1b91a5060a4a9cdd65 && \
-    export TRILINOS_SHORT_HASH=4c8adc9 && \
+RUN export TRILINOS_HASH=99b0900c648ac0b154ba1b06bc5b1d394d6421df && \
+    export TRILINOS_SHORT_HASH=99b0900 && \
     export TRILINOS_URL=https://github.com/trilinos/Trilinos/archive/${TRILINOS_HASH}.tar.gz && \
     export TRILINOS_ARCHIVE=${PREFIX}/archive/trilinos-${TRILINOS_HASH}.tar.xz && \
```
Comment (dalg24) [inline docker/Dockerfile_stack]: .tar.gz

File: docker/Dockerfile_stack [hunk 2]
```diff
@@ -74,22 +70,48 @@ RUN export OPENMP_VERSION=4.0.0 && \
...
+    tar -xf ${OPENMPI_ARCHIVE} -C ${OPENMPI_SOURCE_DIR} --strip-components=1 && \
+    mkdir -p ${OPENMPI_BUILD_DIR} && \
+    cd ${OPENMPI_BUILD_DIR} && \
+    ${OPENMPI_SOURCE_DIR}/configure --prefix=${OPENMPI_INSTALL_DIR} && \
+    make -j${NPROC} install && \
+    rm -rf ${OPENMPI_ARCHIVE} && \
+    rm -rf ${OPENMPI_BUILD_DIR} && \
+    rm -rf ${OPENMPI_SOURCE_DIR}
+
+ENV OPENMPI_DIR=/opt/openmpi/3.0/bin
```
Comment (dalg24) [inline docker/Dockerfile_stack]: To be more consistent with other `_DIR` env variables we set I would do `OPENMPI_DIR=/opt/openmpi/3.0`

File: docker/Dockerfile_stack [hunk 3]
```diff
@@ -74,22 +70,48 @@ RUN export OPENMP_VERSION=4.0.0 && \
...
+# Put OPENMPI_DIR at the end of the path so that /ust/local/bin/mpiexec will
+# overwrite it
+ENV PATH=$PATH:${OPENMPI_DIR}
```
Comment (dalg24) [inline docker/Dockerfile_stack]: In which case this would become `PATH=$PATH:${OPENMPI_DIR}/bin`

File: docker/Dockerfile_stack [hunk 4]
```diff
@@ -180,9 +202,9 @@ RUN export PETSC_VERSION=3.6.4 && \
     ${PETSC_SOURCE_DIR}/configure \
         --prefix=${PETSC_INSTALL_DIR} \
         --with-shared-libraries \
-        --with-cc=/usr/bin/mpicc \
-        --with-cxx=/usr/bin/mpicxx \
-        --with-fc=/usr/bin/mpifort \
+        --with-cc=${OPENMPI_DIR}/mpicc \
+        --with-cxx=${OPENMPI_DIR}/mpicxx \
+        --with-fc=${OPENMPI_DIR}/mpifort \
```
Comment (dalg24) [inline docker/Dockerfile_stack]: `${OPENMPI_DIR}/bin/...`

File: docker/Dockerfile_stack [hunk 5]
```diff
@@ -267,7 +289,7 @@ ENV KOKKOS_TOOLS_DIR=/scratch/source/kokkos-tools/${KOKKOS_TOOLS_SHORT_HASH}/src
 
 # append the option flag --allow-run-as-root to mpiexec
 RUN echo '#!/usr/bin/env bash' > /usr/local/bin/mpiexec && \
-    echo '/usr/bin/mpiexec --allow-run-as-root "$@"' >> /usr/local/bin/mpiexec && \
+    echo '${OPENMPI_DIR}/mpiexec --allow-run-as-root "$@"' >> /usr/local/bin/mpiexec && \
```
Comment (dalg24) [inline docker/Dockerfile_stack]: Same `${OPENMPI_DIR}/bin/mpiexec`

**Comment by dalg24** [comment]
This is ready for review/merge.  The images have already been tagged and pushed to Docker Hub.

File: docker/Dockerfile_stack [hunk 6]
```diff
@@ -253,7 +250,7 @@ RUN ["/bin/bash","-c","export KOKKOS_TOOLS_HASH=7ee5bcb80a8329c8493da735bc7bad5a
                                \"space-time-stack\" \
                               ) && \
     export KOKKOS_TOOLS_URL=https://github.com/kokkos/kokkos-tools/archive/${KOKKOS_TOOLS_HASH}.tar.gz && \
-    export KOKKOS_TOOLS_ARCHIVE=${PREFIX}/archive/kokkos-tools-${KOKKOS_TOOLS_HASH}.tar.xz && \
+    export KOKKOS_TOOLS_ARCHIVE=${PREFIX}/archive/kokkos-tools-${KOKKOS_TOOLS_HASH}.tar.gz && \
```
Comment (aprokop) [inline docker/Dockerfile_stack]: Is `xz` not available? If it is, why the change?

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 7]
```diff
@@ -25,8 +25,10 @@ std::vector<std::array<double, 3>>
 make_stuctured_cloud( double Lx, double Ly, double Lz, int nx, int ny, int nz )
 {
     std::vector<std::array<double, 3>> cloud( nx * ny * nz );
-    std::function<int( int, int, int )> ind = [nx, ny, nz](
-        int i, int j, int k ) { return i + j * nx + k * ( nx * ny ); };
+    std::function<int( int, int, int )> ind = [nx, ny, nz]( int i, int j,
+                                                            int k ) {
```
Comment (aprokop) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: Oh boy, this is quite ugly...

File: packages/Search/src/DTK_LinearBVH_def.hpp [hunk 8]
```diff
@@ -51,9 +51,9 @@ class SetBoundingBoxesFunctor
 template <typename DeviceType>
 BVH<DeviceType>::BVH( Kokkos::View<Box const *, DeviceType> bounding_boxes )
     : _leaf_nodes( "leaf_nodes", bounding_boxes.extent( 0 ) )
-    , _internal_nodes(
-          "internal_nodes",
-          bounding_boxes.extent( 0 ) > 0 ? bounding_boxes.extent( 0 ) - 1 : 0 )
+    , _internal_nodes( "internal_nodes", bounding_boxes.extent( 0 ) > 0
+                                             ? bounding_boxes.extent( 0 ) - 1
```
Comment (aprokop) [inline packages/Search/src/DTK_LinearBVH_def.hpp]: I liked the previous one better...

File: docker/Dockerfile_stack (hunk 6 above)
Comment (dalg24) [reply docker/Dockerfile_stack]: GitHub let you download either a tarball (.tar.gz) or a zipball (.zip) archive.  See [here](https://developer.github.com/v3/repos/contents/#get-archive-link).  We were just changing the extension name by mistake.

**Comment by dalg24** [comment]
@aprokop I have no opinion on the format.  Some changes look good to me, some not so good.  The format configuration is quite arbitrary, you are welcome to do try some changes in the `.clang-format` configuration file if you like.

File: cmake/CodeFormat.cmake [hunk 9]
```diff
@@ -1,8 +1,8 @@
 if(NOT CLANG_FORMAT_EXECUTABLE)
     find_program(CLANG_FORMAT_EXECUTABLE
         NAMES
-        clang-format-4.0
-        clang-format-mp-4.0
+        clang-format-5.0
+        clang-format-5.0
```
Comment (sslattery) [inline cmake/CodeFormat.cmake]: should this be here twice? is this supposed to be clang-format-mp-5.0?

File: docker/Dockerfile_stack [hunk 10]
```diff
@@ -47,12 +47,7 @@ RUN apt-get update && apt-get install -y \
     rm -rf /var/lib/apt/lists/*
 
 RUN pip install --upgrade --no-cache-dir pip && \
-    pip install --no-cache-dir sphinx==1.4 sphinx_rtd_theme breathe==4.6 && \
-    echo "deb http://apt.llvm.org/xenial/ llvm-toolchain-xenial-4.0 main" > /etc/apt/sources.list.d/llvm.list && \
```
Comment (sslattery) [inline docker/Dockerfile_stack]: So we dont have to do all this for 5.0 anymore?

File: docker/Dockerfile_stack (hunk 6 above)
Comment (sslattery) [reply docker/Dockerfile_stack]: didnt know there was an xz

File: docker/Dockerfile_stack [hunk 11]
```diff
@@ -238,10 +235,10 @@ RUN export TRILINOS_HASH=4c8adc92fcbad83b827f9f1b91a5060a4a9cdd65 && \
 ENV TRILINOS_DIR=/scratch/source/trilinos/release
 
 
-# download Kokkos-Tools
+# download Kokkos Profiling and Debugging Tools
 # sh does not support arrays so we need to use bash
-ENV KOKKOS_TOOLS_SHORT_HASH=7ee5bcb
-RUN ["/bin/bash","-c","export KOKKOS_TOOLS_HASH=7ee5bcb80a8329c8493da735bc7bad5a859dfb1f && \
+ENV KOKKOS_TOOLS_SHORT_HASH=2d75b69
```
Comment (sslattery) [inline docker/Dockerfile_stack]: so we are pulling in this specific hash of Kokkos-Tools?

File: docs/source/coding_guidelines.rst [hunk 12]
```diff
@@ -8,7 +8,7 @@ the repository.
 ClangFormat
 -----------
 
-ClangFormat (version 4.0) is used to check the C++ code formatting style in
+ClangFormat (version 5.0) is used to check the C++ code formatting style in
```
Comment (sslattery) [inline docs/source/coding_guidelines.rst]: should we even list the version here?

File: packages/Search/examples/bvh_driver/bvh_driver.cpp (hunk 7 above)
Comment (sslattery) [reply packages/Search/examples/bvh_driver/bvh_driver.cpp]: so these are formatting changes from moving from 4.0 to 5.0?

File: packages/Search/test/tstLinearBVH.cpp [hunk 13]
```diff
@@ -11,7 +11,7 @@
 
 #include <Teuchos_UnitTestHarness.hpp>
 
-#include <boost/geometry/index/rtree.hpp>
+#include <boost/geometry.hpp>
```
Comment (sslattery) [inline packages/Search/test/tstLinearBVH.cpp]: did they change the header install structure in 1.56?

File: cmake/CodeFormat.cmake (hunk 9 above)
Comment (dalg24) [reply cmake/CodeFormat.cmake]: Good catch

File: docker/Dockerfile_stack (hunk 10 above)
Comment (dalg24) [reply docker/Dockerfile_stack]: This was old copy/paste from Dockerfile_clang-format and the LLVM part ended being duplicated in this file.

File: docker/Dockerfile_stack (hunk 11 above)
Comment (dalg24) [reply docker/Dockerfile_stack]: Yes.  It is the current HEAD.

File: docs/source/coding_guidelines.rst (hunk 12 above)
Comment (dalg24) [reply docs/source/coding_guidelines.rst]: I would think so.  Because the formatting end up (sigh) depending on the specific version.

File: packages/Search/test/tstLinearBVH.cpp (hunk 13 above)
Comment (dalg24) [reply packages/Search/test/tstLinearBVH.cpp]: It looks like they did but `<boost/geometry.hpp>` will work with older versions so I think that's the way to go.

File: docker/Dockerfile_stack (hunk 6 above)
Comment (dalg24) [reply docker/Dockerfile_stack]: It does exist.

Comment (dalg24) [reply docker/Dockerfile_stack]: Not on GH though.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp (hunk 7 above)
Comment (dalg24) [reply packages/Search/examples/bvh_driver/bvh_driver.cpp]: Yes.  All the changes to the C++ code in this PR correspond to applying the 5.0 new formatting style.
